### PR TITLE
[Merged by Bors] - feat: add mod n cyclotomic character

### PR DESCRIPTION
**Pull request by kbuzzard**
We define the group homomorphism Aut(L)->(Z/dZ)^* coming from the action on the n'th roots of unity in L. Here d is the number of n'th roots of unity in L (which might not be n in this generality).

We also make API for the standard special case when there are n n'th roots of unity in L, calling it `ModularCyclotomicCharacter`.

Joint work with Hanneke Wiersema, coming from the Leiden conference on machine checked mathematics.

---

- [x] depends on: #11075 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kbuzzard** [comment]
Eric Rodriguez has pointed out the related declaration https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/RootsOfUnity/Basic.html#IsPrimitiveRoot.autToPow  which probably means that I'd like another think about this PR (autToPow is weaker than what we have here but perhaps I can use it to make some proofs shorter?)

**Comment by ericrbg** [comment]
the whole file works just as well with `CommRing` + `IsDomain`. Should this not be going to some units group morally? Not that I know the nice way to implement this. I guess the only plug in that can be done with `autToPow` is for the case where the primitive root exists (I see why on Zulip you said this is fundamentally different, Kevin)

**Comment by kbuzzard** [comment]
Thanks Eric: I changed field to integral domain. I don't like autToPow because it depends on the choice of a primitive root. My instinct is to replace autToPow by my definition and weaken the hypotheses to just the statement that a primitive root exists (in a future PR).

**Comment by ericrbg** [comment]
Yes, I think that's a good idea. I think there's also some possibility that it's nice to have this instead be an equiv to `ZMod d`^x, with proving that d is equal to the size of the group of roots of unity instead. I think it'd give nicer definitional equalities for the autToPow use-case (gal group of cycl extension) and if it's unwanted then we can just use `rfl`

**Review by riccardobrasca — COMMENTED** [review]
I am wondering whether it makes sense to connect this with [MulChar](https://leanprover-community.github.io/mathlib4_docs/Mathlib/NumberTheory/LegendreSymbol/MulCharacter.html#MulChar). Maybe it's very relevant, since the main point of `MulChar` is to have a ring as domain. @MichaelStollBayreuth what do you think?

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+/-- If `g` is a field automorphism of `L`, and `n : ℕ+`, then
+  `ModularCyclotomicCharacter.toFun n g` is the `j : ZMod d` such that `g(ζ)=ζ^j` for all
+  `n`'th roots of unity. Here `d` is the number of `n`th roots of unity in `L`. -/
+noncomputable def ModularCyclotomicCharacter.toFun (n : ℕ+) (g : L ≃+* L) :
+    ZMod (Fintype.card (rootsOfUnity n L)) :=
+  ModularCyclotomicCharacter_aux g n
+
+-- This appears to be missing from the library. It should not be in this file.
+theorem Group.pow_eq_zpow_mod {G : Type _} [Group G] {x : G} (m : ℤ) {n : ℕ} (h : x ^ n = 1) :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:101]:
```suggestion
theorem Group.pow_eq_zpow_mod {G : Type*} [Group G] {x : G} (m : ℤ) {n : ℕ} (h : x ^ n = 1) :
```

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+theorem spec' (g : L ≃+* L) {n : ℕ+} {t : Lˣ} (ht : t ∈ rootsOfUnity n L) :
+    g t = t ^ (χ n g).val :=
+  spec g ⟨t, ht⟩
+
+theorem spec'' (g : L ≃+* L) {n : ℕ+} {t : L} (ht : IsPrimitiveRoot t n) :
+    g t = t ^ (χ n g).val :=
+  spec' g (SetLike.coe_mem ht.toRootsOfUnity)
+
+-- this is in the wrong place I guess
+lemma ext {G : Type _} [Group G] [Fintype G] [IsCyclic G]
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:127]:
```suggestion
lemma ext {G : Type*} [Group G] [Fintype G] [IsCyclic G]
```

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  { toFun := ModularCyclotomicCharacter.toFun n
+    map_one' := ModularCyclotomicCharacter.id n
+    map_mul' := ModularCyclotomicCharacter.comp n }
+
+/-- Given a positive integer `n` and a field `L` containing `n` `n`th roots
+of unity, `ModularCyclotomicCharacter n` is a multiplicative homomorphism from the
+automorphisms of `L` to `ℤ/nℤ`. It is uniquely characterised by the property that
+`g(ζ)=ζ^(ModularCyclotomicCharacter n g)` for `g` an automorphism of `L` and `ζ` any `n`th root
+of unity. -/
+noncomputable def ModularCyclotomicCharacter (n : ℕ+)
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:183]: `n` could be `Nat` and implicit here (the cardinality in `hn` cannot be zero).

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+noncomputable def ModularCyclotomicCharacter.toFun (n : ℕ+) (g : L ≃+* L) :
+    ZMod (Fintype.card (rootsOfUnity n L)) :=
+  ModularCyclotomicCharacter_aux g n
+
+-- This appears to be missing from the library. It should not be in this file.
+theorem Group.pow_eq_zpow_mod {G : Type _} [Group G] {x : G} (m : ℤ) {n : ℕ} (h : x ^ n = 1) :
+    x ^ m = x ^ (m % (n : ℤ)) := by
+  have h2 : x ^ m = x ^ ((n : ℤ) * (m / (n : ℤ)) + m % (n : ℤ)) :=
+    congr_arg (fun a => x ^ a) ((Int.add_comm _ _).trans (Int.emod_add_ediv _ _)).symm
+  simp [h, h2, zpow_add, zpow_mul]
```
[the remark this answers is not in the document]
Comment (kbuzzard) [reply Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:105]: Nice! I've split this off as #11009 . Sorry, I opened this PR but forgot to move those a couple of proofs to where they were supposed to be beforehand!

[the remark this answers is not in the document]
Comment (kbuzzard) [reply Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean]: #11009 is now merged.

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (kbuzzard) [reply Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean]: I've put both lemmas in #11016

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean (hunk 3 above)
Comment (kbuzzard) [reply Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:164]: Unfortunately `rootsOfUnity n` needs `n : ℕ+` so I don't see how to make this work.

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean (hunk 2 above)
Comment (kbuzzard) [reply Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean]: Fixed in #11016

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean (hunk 1 above)
Comment (kbuzzard) [reply Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean]: Fixed in #11009

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+namespace ModularCyclotomicCharacter
+
+local notation "χ" => ModularCyclotomicCharacter.toFun
+
+/-- The formula which characterises the output of `ModularCyclotomicCharacter g n`. -/
+theorem spec (g : L ≃+* L) {n : ℕ+} (t : rootsOfUnity n L) :
```
[the remark this answers is not in the document]
Comment (kbuzzard) [reply Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:104]: Now added as `ModularCyclotomicCharacter.unique` (and indeed I could golf `id` and `comp`).

**Comment by kbuzzard** [comment]
> This depends on your other PR, right? Can you mark it in the description? Thanks!

OK! I had just left it tagged awaiting-author and was waiting for the other PR to be merged, but if you think it's OK to just make it a dependent PR I can tag it awaiting-review now (I am never quite sure of the etiquette here).

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+* `ModularCyclotomicCharacter L n hn : (L ≃+* L) →* (ZMod n)ˣ` sends `g` to the unique `j` such
+   that `g(ζ)=ζ^j` for all `ζ : rootsOfUnity n L`. Here `hn` is a proof that there
+   are `n` `n`th roots of unity in `L`.
+
+## Implementation note
+
+In theory this could be set up as some theory about monoids, being a character
+on monoid isomorphisms, but under the hypotheses that the `n`'th roots of unity
+are cyclic of order `n`. The advantage of sticking to integral domains is that finite subgroups
```
Comment (kbuzzard) [inline Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:38]:
```suggestion
are cyclic. The advantage of sticking to integral domains is that finite subgroups
```

File: Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+field anyway.
+
+Although I don't know whether it's of any use, `ModularCyclotomicCharacter'`
+is the general case for integral domains, with target in `(ZMod d)ˣ`
+where `d` is the number of `n`th roots of unity in `L`.
+
+## Todo
+
+* Prove the compatibility of `ModularCyclotomicCharacter n` and `ModularCyclotomicCharacter m`
+if `n ∣ m`.
```
Comment (kbuzzard) [inline Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean:50]:
```suggestion
  if `n ∣ m`.
```

**Comment by faenuccio** [comment]
In your overall set-up, your field $L$ is fixed and you consider the group of its (ring) automorphisms as domain for $\chi_0$. This prevents you from considering the cyclotomic character relative to a fixed "base" field in the following sense: if $F/K$ is an extension of number fields (say...) both embedded in $L=F(\zeta_{p^{\infty}})$ then you only get an "absolute" cyclotomic character rather than one, $\chi_K$, relative to $K$ and one, $\chi_F$ relative to $F$ whose image will possibly be a proper subgroup of the image of $\chi_K$. Now, it seems to me that it is useful to keep track of this information and with your construction one would constantly need to change $\chi_0$ to a suitable power to get a new character whose image is the one you are interested in.

I wonder if the construction should not rather be parametrised over an algebra $L/K$ and the domain should be the $K$-algebra isomorphisms of $L$. This would make it more compatible with the definition of the absolute Galois group we have now in Mathlib.

**Comment by kbuzzard** [comment]
@faenuccio There's a canonical forgetful group homomorphism from Gal(L/K) to the group of field automorphisms of L, and all I have to do to get your vision of the modular cyclotomic character is to just compose with this map. Are you suggesting I made this composition into a new definition?

**Comment by faenuccio** [comment]
> @faenuccio There's a canonical forgetful group homomorphism from Gal(L/K) to the group of field automorphisms of L, and all I have to do to get your vision of the modular cyclotomic character is to just compose with this map. Are you suggesting I made this composition into a new definition?

Yes, I am suggesting that you create also a "relative" cyclotomic character, either by doing as you suggest or by defining it as a group hom from $\mathrm{Aut}_K(L)\to (\mathbb{Z}/d)^\times$, and prove some results like $\chi_K^a=\chi_L$ where $a$ is the index of $\mu_K$ inside $\mu_L$. I think that this gadget is the one that really shows off more often e.g. in interpolating special values of $L$-functions beyond Dirichlet ones. What troubles me with this "absolute" definition is that if I am working over $\mathbb{Q}(\zeta_p)$ I would like the image of the cyclotomic character to be pro- $p$ , for instance. 

Lean-wise, if you will want to twist a Galois rep of $G_K$ by a power of it, you will constantly need to compose the absolute one with the group hom you speak about, and this can be painful especially when letting $K$ vary.

**Comment by riccardobrasca** [comment]
> > @faenuccio There's a canonical forgetful group homomorphism from Gal(L/K) to the group of field automorphisms of L, and all I have to do to get your vision of the modular cyclotomic character is to just compose with this map. Are you suggesting I made this composition into a new definition?
> 
> Yes, I am suggesting that you create also a "relative" cyclotomic character, either by doing as you suggest or by defining it as a group hom from AutK(L)→(Z/d)×, and prove some results like χKa=χL where a is the index of μK inside μL. I think that this gadget is the one that really shows off more often e.g. in interpolating special values of L-functions beyond Dirichlet ones. What troubles me with this "absolute" definition is that if I am working over Q(ζp) I would like the image of the cyclotomic character to be pro- p , for instance.
> 
> Lean-wise, if you will want to twist a Galois rep of GK by a power of it, you will constantly need to compose the absolute one with the group hom you speak about, and this can be painful especially when letting K vary.

Do you think we need to replace this one with your version, or we want both?

**Comment by faenuccio** [comment]
> > > @faenuccio There's a canonical forgetful group homomorphism from Gal(L/K) to the group of field automorphisms of L, and all I have to do to get your vision of the modular cyclotomic character is to just compose with this map. Are you suggesting I made this composition into a new definition?
> > 
> > 
> > Yes, I am suggesting that you create also a "relative" cyclotomic character, either by doing as you suggest or by defining it as a group hom from AutK(L)→(Z/d)×, and prove some results like χKa=χL where a is the index of μK inside μL. I think that this gadget is the one that really shows off more often e.g. in interpolating special values of L-functions beyond Dirichlet ones. What troubles me with this "absolute" definition is that if I am working over Q(ζp) I would like the image of the cyclotomic character to be pro- p , for instance.
> > Lean-wise, if you will want to twist a Galois rep of GK by a power of it, you will constantly need to compose the absolute one with the group hom you speak about, and this can be painful especially when letting K vary.
> 
> Do you think we need to replace this one with your version, or we want both?

I think I would keep both. On a one hand, this one can be useful to construct the other one and on the other hand one might want to simply use the map leaving from $\mathrm{Aut}(L)$ as a field. Probably, though, I would call my version the cyclotomic "character" and this one something else (cyclotomic hom?).

**Comment by riccardobrasca** [comment]
I think we merge this and generalize it later, thanks!

bors merge

**Pull request closed without merging**